### ISSUE TITLE
Having problem with responding to a POST request in OnBeforeRequest event

### ISSUE THREAD

**Issue by noldoreldar** [justcoding121/Titanium-Web-Proxy#513]
I don't know if I'm doing something that I shouldn't do, or if I am using a very bad practice here.. But I'm making a project that the proxy server should generate its own custom response. 

Yet I couldn't manage to return a successful response to **POST requests that have a body** content. 

Here are some explanation, some code and the some sample HTTP request.

-- I'm testing this with Firefox browser, changed its proxy setting to 9933 (my explicit endpoint).. 

-- I added the certificate which is generated by the code into Firefox exceptions.. 

-- When I visit a web site and trace Network activity on Firefox, I see all my GET requests are being responded perfectly, but POST requests are just waiting on the air. 

-- This doesn't change even if I send the POST request on its own without visiting a website.

-- When I debug the code, there are no errors or whatsoever. The code just works the same way GET requests are working, passes all the code lines to the end. But I can't see any result on Firefox. 

-- I have also tried putting Fiddler in the middle to see what is going on. Fiddler also behaves the same way Firefox does, GET requests are fine, POST requests are just waiting without any response.

-- I thought maybe it was because I didn't configure any HTTPS certificate options, so I changed the code to use Fiddler's Root Certifcate.. The result didn't change..

-- I thought it could be something related to the request headers.. So i removed  "keep-alive", "Upgrade-Insecure-Requests" and "Accept-Encoding" headers and sent very basic headers only. Nothing changed..

> 
> **EDIT :** 
> After a couple of retries, it seems like I was supposed to do this custom response thing in "OnBeforeResponse" event instead. Yet I am not sure what is going on in the proxy server internal code between "OnBeforeRequest" and "OnBeforeResponse" events. I need to reply to the client user without making any actual data request from the requested URL.. If proxy server itself fetches any data before "OnBeforeRequest" from the requested URL, I need another workaround..

```
using System;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using Titanium.Web.Proxy;
using Titanium.Web.Proxy.EventArguments;
using Titanium.Web.Proxy.Http;
using Titanium.Web.Proxy.Models;
```

```
namespace TitaniumRespondTest
{
    class Program
    {

        static void Main(string[] args)
        {
            var proxyServer = new ProxyServer();
            proxyServer.BeforeRequest += _onBeforeRequestAsync;
            var transparentEndPoint = new TransparentProxyEndPoint(IPAddress.Parse("127.0.0.1"), 9934, false);
            proxyServer.AddEndPoint(transparentEndPoint);
            var explicitEndPoint = new ExplicitProxyEndPoint(IPAddress.Parse("127.0.0.1"), 9933);
            proxyServer.AddEndPoint(explicitEndPoint);
            proxyServer.Start();
            Console.Read();
        }

        private static async Task _onBeforeRequestAsync(object sender, SessionEventArgs e)
        {
            try
            {
                if (e.WebSession.Request.HasBody)
                {
                    byte[] bodyBytes = await e.GetRequestBody();
                    e.SetRequestBody(bodyBytes);
                    e.UserData = e.WebSession.Request;
                }
                var response = await CreateCustomResponse(e.WebSession.Request);
                e.Respond(response);
                if (response.HasBody)
                {
                    e.SetResponseBody(response.Body);
                }
            }
            catch (Exception ex)
            {
                // Error LOG : _log.Error("Error while running request on gecko controller.", ex);
                var response = new Response(Encoding.UTF8.GetBytes($"Error : {ex}"))
                {
                    HttpVersion = e.WebSession.Request.HttpVersion,
                    StatusCode = 500,
                    StatusDescription = ex.Message,
                    ContentType = "text/plain"
                };
                e.Respond(response, true);
                e.SetResponseBody(response.Body);
            }
        }

        public static Task<Response> CreateCustomResponse(Request request)
        {
            // Originally there are some private code here that I generate a response programmatically..
            // here I will only generate a very simple response, yet still it will not work
            var htmlStr = @"<!DOCTYPE html>
                            <html lang=""en-US"">
                            <head>
                              <meta charset=""utf-8"">   
                              <title>TEST</title>
                              <meta name=""viewport"" content=""width=device-width, initial-scale=1"">   
                            </head>
                            <body><div>TEST</div></body>
                            </html>";
            var response = new Response(Encoding.UTF8.GetBytes(htmlStr));
            response.HttpVersion = request.HttpVersion;
            response.StatusCode = 200;
            response.StatusDescription = "OK";
            response.Headers.AddHeader("content-type", "text/html; charset=utf-8");
            return Task.FromResult(response);
        }
   }
}
```

```
POST https://**betterhidedomainname**/register HTTP/1.1
Host: https://**betterhidedomainname**
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Content-Type: application/x-www-form-urlencoded
Content-Length: 15
Connection: keep-alive
Upgrade-Insecure-Requests: 1

whaaa=&asdasd=1
```

**Comment by justcoding121** [comment]
If you respond inside BeforeRequest it will not make a request to server. But BeforeResponse will do.

**Comment by noldoreldar** [comment]
> If you respond inside BeforeRequest it will not make a request to server. But BeforeResponse will do.

As i mentioned, responding inside BeforeRequest only worked for GET requests for me. Responding to POST requests somehow fails, in a way like TCP connection stays open and client never receives the result. This was fixed when I moved my code into BeforeRespond event. I don't know if this is a bug, or it's how it supposed to work..

Are you certainly sure that ProxyServer object fetches data from target URL on its own after BeforeRequest??
